src/librustc_resolve/lib.rs

```diff
@@ -4579,37 +4580,35 @@ impl<'a> Resolver<'a> {
                     false,
                     module.span,
                 ).is_ok() {
-                    let import_id = match binding.kind {
-                        NameBindingKind::Import { directive, .. } => {
-                            self.maybe_unused_trait_imports.insert(directive.id);
-                            self.add_to_glob_map(&directive, trait_name);
-                            Some(directive.id)
-                        }
-                        _ => None,
-                    };
+                    let import_ids = self.find_transitive_imports(&binding.kind, &trait_name);
                     let trait_def_id = module.def_id().unwrap();
-                    found_traits.push(TraitCandidate { def_id: trait_def_id, import_id });
+                    found_traits.push(TraitCandidate { def_id: trait_def_id, import_ids });
                 }
             } else if let Def::TraitAlias(_) = binding.def() {
                 // For now, just treat all trait aliases as possible candidates, since we don't
                 // know if the ident is somewhere in the transitive bounds.
-
-                let import_id = match binding.kind {
-                    NameBindingKind::Import { directive, .. } => {
-                        self.maybe_unused_trait_imports.insert(directive.id);
-                        self.add_to_glob_map(&directive, trait_name);
-                        Some(directive.id)
-                    }
-                    _ => None,
-                };
-                let trait_def_id = binding.def().def_id();
-                found_traits.push(TraitCandidate { def_id: trait_def_id, import_id });
+                let import_ids = self.find_transitive_imports(&binding.kind, &trait_name);
+                let trait_def_id = module.def_id().unwrap();
+                found_traits.push(TraitCandidate { def_id: trait_def_id, import_ids });
             } else {
                 bug!("candidate is not trait or trait alias?")
             }
         }
     }
 
+    fn find_transitive_imports(&mut self, kind: &NameBindingKind<'_>,
+                               trait_name: &Ident) -> SmallNodeIdVec {
+        let mut import_ids = smallvec![];
+        let mut kind = kind;
```

Comment:
Nit: you can make the function parameter `mut` instead.

Reply:
Done